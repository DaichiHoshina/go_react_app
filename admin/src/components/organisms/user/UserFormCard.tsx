import { Card } from "@material-ui/core";
import { useRouter } from "next/router";
import { useSnackbar } from "notistack";
import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { TUser, TUserState } from "../../../modules/User";
import { createUser, fetchUser, updateUser } from "../../../services/User";
import CreateOrEditButton from "../../atoms/share/CreateOrEditButton";
import SeparateHr from "../../atoms/share/SeparateHr";
import TextFieldParts from "../../atoms/share/TextFieldParts";
import { useFormik } from "formik";
import KeyValuePair from "../common/KeyValuePair";

interface Props {
  isEditPage: boolean;
  accountInfo?: any;
  isUserSettingPage?: boolean;
}

const UserFormCard: React.FC<Props> = ({
  isEditPage = false,
  isUserSettingPage = false,
}) => {
  const router = useRouter();
  const { id } = router.query;
  const { enqueueSnackbar } = useSnackbar();
  const dispatch = useDispatch();
  const state = useSelector((state: { userState: TUserState }) => state);

  useEffect(() => {
    dispatch(fetchUser({ id: id }));
  }, [router.query]);

  useEffect(() => {
    if (state?.userState?.user) {
      formik.setValues(state?.userState?.user!);
    }
  }, [state?.userState?.user]);

  const formik = useFormik<TUser>({
    initialValues: {},
    onSubmit: async (values) => {
      const response = (await isEditPage)
        ? dispatch(updateUser({ user: values, id: id }))
        : dispatch(createUser({ user: values }));
      if (response.arg) {
        enqueueSnackbar(isEditPage ? "Update!!" : "Create!!", {
          variant: "success",
        });
        router.push(isUserSettingPage ? `/users/${id}` : "/users");
      } else {
        enqueueSnackbar("Failure...", {
          variant: "error",
        });
      }
    },
  });

  return (
    <>
      <Card className="p-5 w-4/5">
        <ul className="flex flex-col space-y-2">
          <KeyValuePair
            keyName="氏名"
            value={
              <div className="flex">
                <TextFieldParts name="name" formik={formik} />
              </div>
            }
          />
        </ul>
        <SeparateHr />
        <div className="flex justify-end">
          <CreateOrEditButton isEditPage={isEditPage} formik={formik} />
        </div>
      </Card>
    </>
  );
};

export default UserFormCard;
