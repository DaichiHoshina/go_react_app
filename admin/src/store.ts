import { combineReducers } from 'redux';
import logger from 'redux-logger';
import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { userSlice, initialState as userInitialState } from './modules/User';

const rootReducer = combineReducers({
  userState: userSlice.reducer,
});

const preloadedState = () => {
  return { userState: userInitialState };
};

const store = () => {
  const middlewareList = [...getDefaultMiddleware(), logger];

  return configureStore({
    reducer: rootReducer,
    middleware: middlewareList,
    devTools: process.env.NODE_ENV !== 'production',
    preloadedState: preloadedState(),
  });
};

export default store;
