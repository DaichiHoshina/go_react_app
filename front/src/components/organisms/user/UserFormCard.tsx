import { Card } from "@material-ui/core";
import { useRouter } from "next/router";
import { useSnackbar } from "notistack";
import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { TUser, TUserState } from "../../../modules/User";
import { createUser, fetchUser, updateUser } from "../../../services/User";
import CreateOrEditButton from "../../atoms/share/CreateOrEditButton";
import SeparateHr from "../../atoms/share/SeparateHr";
import TextFieldParts from "../../atoms/share/TextFieldParts";
import { useFormik } from "formik";
import KeyValuePair from "../common/KeyValuePair";
import { UserSettingUpdateSchema } from "../../../const/validation";
import { DropzoneArea } from "material-ui-dropzone";
import { fetchPresentations } from "../../../services/Presentation";

interface Props {
  isEditPage: boolean;
  accountInfo?: any;
  isUserSettingPage?: boolean;
}

const UserFormCard: React.FC<Props> = ({
  isEditPage = false,
  isUserSettingPage = false,
}) => {
  const router = useRouter();
  const dispatch = useDispatch();
  const state = useSelector((state: { userState: TUserState }) => state);
  const { id } = router?.query || {};
  const { enqueueSnackbar } = useSnackbar() || {};

  useEffect(() => {
    dispatch(fetchUser({ id: id }));
  }, [id]);

  useEffect(() => {
    if (state?.userState?.user) {
      formik.setValues(state?.userState?.user!);
    }
  }, [state?.userState?.user]);

  const formik = useFormik<TUser>({
    initialValues: {},
    validationSchema: UserSettingUpdateSchema,
    onSubmit: async (values) => {
      const response: any = (await isEditPage)
        ? dispatch(updateUser({ user: values, id: id }))
        : dispatch(createUser({ user: values }));
      if (response.arg) {
        enqueueSnackbar("データを送信中", {
          variant: "success",
        });
        // 画像が保存されるまでタイムラグがあるため、○秒後に実行するようにしている
        setTimeout(function () {
          enqueueSnackbar(isEditPage ? "Update!!" : "Create!!", {
            variant: "success",
          });
          dispatch(
            fetchPresentations({
              page: 1,
              per: 1,
            })
          );
          router.push(isUserSettingPage ? `/users/${id}` : "/presentations");
        }, 2000);
      } else {
        enqueueSnackbar("Failure...", {
          variant: "error",
        });
      }
    },
  });

  return (
    <>
      <Card className="p-5 w-4/5">
        <ul className="flex flex-col space-y-2">
          <div className="mb-3">
            <KeyValuePair
              keyName="image"
              value={
                <>
                  <DropzoneArea
                    dropzoneText={
                      "ここにファイルをドロップ\nまたはファイルを選択"
                    }
                    showPreviews={true}
                    showPreviewsInDropzone={false}
                    showFileNamesInPreview={true}
                    getFileAddedMessage={(fileName: string) =>
                      `${fileName}を選択しました。`
                    }
                    getFileRemovedMessage={(fileName: string) =>
                      `${fileName}を削除しました。`
                    }
                    filesLimit={1}
                    previewText="アップロードファイル"
                    onChange={(files) => {
                      formik.setFieldValue("image", files);
                    }}
                  />
                  {(formik.values?.image ?? []).length === 0 && (
                    <p className="pl-3 pt-2 text-white text-xs">
                      select image.
                    </p>
                  )}
                </>
              }
            />
          </div>
          <KeyValuePair
            keyName="name"
            value={
              <div className="flex">
                <TextFieldParts name="name" formik={formik} />
              </div>
            }
          />
        </ul>
        <SeparateHr />
        <div className="flex justify-end">
          <CreateOrEditButton isEditPage={isEditPage} formik={formik} />
        </div>
      </Card>
    </>
  );
};

export default UserFormCard;
