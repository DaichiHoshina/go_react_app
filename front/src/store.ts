import { combineReducers } from "redux";
import logger from "redux-logger";
import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
import { userSlice, initialState as userInitialState } from "./modules/User";
import {
  presentationSlice,
  initialState as presentationInitialState,
} from "./modules/Presentation";

const rootReducer = combineReducers({
  userState: userSlice.reducer,
  presentationState: presentationSlice.reducer,
});

const preloadedState = () => {
  return {
    userState: userInitialState,
    presentationState: presentationInitialState,
  };
};

const store = () => {
  const middlewareList = [...getDefaultMiddleware(), logger];

  return configureStore({
    reducer: rootReducer,
    middleware: middlewareList,
    devTools: process.env.NODE_ENV !== "production",
    preloadedState: preloadedState(),
  });
};

export default store;
