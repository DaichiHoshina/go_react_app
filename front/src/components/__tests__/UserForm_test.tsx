import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import store from "../../store";
import UserFormCard from "../organisms/user/UserFormCard";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

it("<UserFormCard>の編集画面スナップショット", () => {
  const tree = renderer
    .create(
      <Provider store={store()}>
        <UserFormCard isEditPage={true} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("投稿編集ページのレンダリング", () => {
  // const handleOpen = jest.fn();

  const { getByText } = render(
    <Provider store={store()}>
      <UserFormCard isEditPage={true} />
    </Provider>
  );

  // inputのテスト
  const input = screen.getByText("name");
  userEvent.type(input, "test");

  // ボタンのテスト
  const submitButton = getByText("UPDATE!");
  expect(submitButton).toBeInTheDocument();
});
