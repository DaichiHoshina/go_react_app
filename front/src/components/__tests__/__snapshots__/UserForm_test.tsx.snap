// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserFormCard>の編集画面スナップショット 1`] = `
<div
  className="MuiPaper-root MuiCard-root p-5 w-4/5 MuiPaper-elevation1 MuiPaper-rounded"
>
  <ul
    className="flex flex-col space-y-2"
  >
    <div
      className="mb-3"
    >
      <li
        className="flex w-full"
      >
        <p
          className="w-1/3"
        >
          image
        </p>
        <div
          className="w-2/3"
        >
          <div
            className="MuiDropzoneArea-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
          >
            <input
              accept=""
              autoComplete="off"
              multiple={false}
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              type="file"
            />
            <div
              className="MuiDropzoneArea-textContainer"
            >
              <p
                className="MuiTypography-root MuiDropzoneArea-text MuiTypography-h5"
              >
                ここにファイルをドロップ
またはファイルを選択
              </p>
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiDropzoneArea-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
                />
              </svg>
            </div>
          </div>
          <p
            className="pl-3 pt-2 text-white text-xs"
          >
            select image.
          </p>
        </div>
      </li>
    </div>
    <li
      className="flex w-full"
    >
      <p
        className="w-1/3"
      >
        name
      </p>
      <div
        className="w-2/3"
      >
        <div
          className="flex"
        >
          <div
            className="w-full"
          >
            <div
              className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  id="outlined-basic"
                  name="name"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder=""
                  required={false}
                  rows={1}
                  style={
                    Object {
                      "padding": 8,
                      "paddingLeft": 10,
                      "paddingRight": 10,
                    }
                  }
                  type="text"
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                  style={
                    Object {
                      "paddingLeft": 8,
                    }
                  }
                >
                  <legend
                    className="PrivateNotchedOutline-legend-2"
                    style={
                      Object {
                        "width": 0.01,
                      }
                    }
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
  <hr
    className="my-5 border-t border-gray-300"
  />
  <div
    className="flex justify-end"
  >
    <div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained mr-5 MuiButton-containedSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          UPDATE!
        </span>
      </button>
    </div>
  </div>
</div>
`;
